File: android/app/src/main/java/com/lowerbackstretching/ui/player/PlayerEngine.kt
package com.lowerbackstretching.ui.player

import com.lowerbackstretching.core.model.Stretch
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharedFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update

/**
 * Pure state machine for the stretch player — no Android dependency, so
 * it's directly unit-testable. The viewmodel drives [tick] on a
 * one-second cadence and listens for the [finishedEvents] flow to record
 * the session exactly once.
 *
 * `startIndex` lets the viewmodel resume an interrupted routine — pass
 * the saved index and the engine starts at that stretch.
 */
class PlayerEngine(
    initialStretches: List<Stretch>,
    startIndex: Int = 0,
) {

    data class Snapshot(
        val stretches: List<Stretch>,
        val index: Int,
        val remainingSeconds: Int,
        val running: Boolean,
        val finished: Boolean,
    ) {
        val current: Stretch? get() = stretches.getOrNull(index)

        val progress: Float
            get() = current?.let {
                if (it.durationSeconds == 0) 0f
                else (it.durationSeconds - remainingSeconds).toFloat() / it.durationSeconds
            } ?: 0f

        /**
         * Progress across the entire routine, weighted by per-stretch
         * duration. 0 at first frame, 1 at the moment the last stretch
         * finishes.
         */
        val routineProgress: Float
            get() {
                if (finished) return 1f
                val total = stretches.sumOf { it.durationSeconds }
                if (total == 0) return 0f
                val elapsedBefore = stretches.take(index).sumOf { it.durationSeconds }
                val elapsedInCurrent = (current?.durationSeconds ?: 0) - remainingSeconds
                return ((elapsedBefore + elapsedInCurrent).toFloat() / total).coerceIn(0f, 1f)
            }
    }

    /** Emitted exactly once when the routine completes. */
    data class FinishedEvent(val totalDurationSeconds: Int)

    private val safeStartIndex = startIndex.coerceIn(0, (initialStretches.size - 1).coerceAtLeast(0))

    private val _state = MutableStateFlow(
        Snapshot(
            stretches = initialStretches,
            index = safeStartIndex,
            remainingSeconds = initialStretches.getOrNull(safeStartIndex)?.durationSeconds ?: 0,
            running = initialStretches.isNotEmpty(),
            finished = initialStretches.isEmpty(),
        )
    )
    val state: StateFlow<Snapshot> = _state.asStateFlow()

    private val _finishedEvents = MutableSharedFlow<FinishedEvent>(extraBufferCapacity = 1)
    val finishedEvents: SharedFlow<FinishedEvent> = _finishedEvents.asSharedFlow()

    val totalDurationSeconds: Int = initialStretches.sumOf { it.durationSeconds }

    /**
     * Advance the clock by one second. Returns true if this tick caused
     * the routine to finish. Callers don't need to use the return value
     * for session recording — subscribe to [finishedEvents] instead.
     */
    fun tick(): Boolean {
        val s = _state.value
        if (!s.running || s.finished) return false
        return if (s.remainingSeconds > 1) {
            _state.update { it.copy(remainingSeconds = it.remainingSeconds - 1) }
            false
        } else {
            advance()
        }
    }

    /** Skip to the next stretch immediately. Returns true if finished. */
    fun next(): Boolean {
        if (_state.value.finished) return false
        return advance()
    }

    fun previous() {
        val s = _state.value
        val prev = (s.index - 1).coerceAtLeast(0)
        _state.update {
            it.copy(
                index = prev,
                remainingSeconds = it.stretches[prev].durationSeconds,
                finished = false,
                running = true,
            )
        }
    }

    fun togglePlay() = _state.update { it.copy(running = !it.running) }

    private fun advance(): Boolean {
        val s = _state.value
        val nextIdx = s.index + 1
        return if (nextIdx >= s.stretches.size) {
            _state.update { it.copy(finished = true, running = false) }
            _finishedEvents.tryEmit(FinishedEvent(totalDurationSeconds))
            true
        } else {
            _state.update {
                it.copy(
                    index = nextIdx,
                    remainingSeconds = it.stretches[nextIdx].durationSeconds,
                )
            }
            false
        }
    }
}

File: ios/LowerBackStretching/Data/PlayerEngine.swift
import Foundation
import Combine

/// Pure state machine for the stretch player — no SwiftUI dependency, so
/// it's directly unit-testable. The view drives `tick()` on a one-second
/// cadence and listens for `finishedEvent` to record the session exactly
/// once.
final class PlayerEngine: ObservableObject {

    struct Snapshot: Equatable {
        var stretches: [Stretch]
        var index: Int
        var remainingSeconds: Int
        var running: Bool
        var finished: Bool

        var current: Stretch? {
            stretches.indices.contains(index) ? stretches[index] : nil
        }

        var progress: Double {
            guard let current, current.durationSeconds > 0 else { return 0 }
            return Double(current.durationSeconds - remainingSeconds) / Double(current.durationSeconds)
        }

        /// Progress across the entire routine, weighted by per-stretch
        /// duration. 0 at first frame, 1 at the moment the last stretch
        /// finishes.
        var routineProgress: Double {
            if finished { return 1 }
            let total = stretches.reduce(0) { $0 + $1.durationSeconds }
            guard total > 0 else { return 0 }
            let elapsedBefore = stretches.prefix(index).reduce(0) { $0 + $1.durationSeconds }
            let elapsedInCurrent = (current?.durationSeconds ?? 0) - remainingSeconds
            return min(1, max(0, Double(elapsedBefore + elapsedInCurrent) / Double(total)))
        }
    }

    /// Set exactly once when the routine completes. Views observe this
    /// (via `.onChange`) to record the session.
    struct FinishedEvent: Equatable {
        let totalDurationSeconds: Int
    }

    @Published private(set) var snapshot: Snapshot
    @Published private(set) var finishedEvent: FinishedEvent?
    let totalDurationSeconds: Int

    /// `startIndex` lets the caller resume an interrupted routine — pass
    /// the saved index and the engine starts at that stretch.
    init(stretches: [Stretch], startIndex: Int = 0) {
        self.totalDurationSeconds = stretches.reduce(0) { $0 + $1.durationSeconds }
        let safe = max(0, min(startIndex, stretches.count - 1))
        let index = stretches.isEmpty ? 0 : safe
        self.snapshot = Snapshot(
            stretches: stretches,
            index: index,
            remainingSeconds: stretches.indices.contains(index)
                ? stretches[index].durationSeconds : 0,
            running: !stretches.isEmpty,
            finished: stretches.isEmpty
        )
    }

    /// Advance the clock by one second. Returns true if this tick caused
    /// the routine to finish. Callers don't need the return value for
    /// session recording — observe `finishedEvent` instead.
    @discardableResult
    func tick() -> Bool {
        guard snapshot.running, !snapshot.finished else { return false }
        if snapshot.remainingSeconds > 1 {
            snapshot.remainingSeconds -= 1
            return false
        } else {
            return advance()
        }
    }

    @discardableResult
    func next() -> Bool {
        guard !snapshot.finished else { return false }
        return advance()
    }

    func previous() {
        let prev = max(0, snapshot.index - 1)
        snapshot.index = prev
        snapshot.remainingSeconds = snapshot.stretches[prev].durationSeconds
        snapshot.finished = false
        snapshot.running = true
    }

    func togglePlay() {
        snapshot.running.toggle()
    }

    private func advance() -> Bool {
        let nextIdx = snapshot.index + 1
        if nextIdx >= snapshot.stretches.count {
            snapshot.finished = true
            snapshot.running = false
            finishedEvent = FinishedEvent(totalDurationSeconds: totalDurationSeconds)
            return true
        } else {
            snapshot.index = nextIdx
            snapshot.remainingSeconds = snapshot.stretches[nextIdx].durationSeconds
            return false
        }
    }
}
